feat(transformer): add `async_generator_functions` plugin

Dunqing opened the pull request:
Passed 15/19 tests. The remaining 4 failed tests related to `this` expression, the problem same as I mentioned in #6658. I will fix them in follow-up PRs.

Dunqing commented:
* **#6573** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/6573?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @Dunqing and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

overlookmotel commented on crates/oxc_transformer/src/es2018/async_generator_functions/for_await.rs:
```diff
@@ -0,0 +1,340 @@
+
+/// Constructs AST based on AstBuilder
+/// <https://github.com/babel/babel/blob/d20b314c14533ab86351ecf6ca6b7296b66a57b3/packages/babel-plugin-transform-async-generator-functions/src/for-await.ts#L3-L30>
+fn build_for_await<'a>(
```
Holy CHRIST! This function is *enormous*. We really do need the `ast!` macro.

Probably you were going to do this anyway, but could you include what code it produces in the comment (written out, rather than just link to Babel's impl)?

overlookmotel replied:
Also, I'm not sure the 3 x `VariableDeclaration`s are actually needed. Doesn't Babel just discard them?

I had assumed they were only in Babel's template to make the `template` function resolve the identifiers correctly (I could be wrong).

Dunqing replied:
I think so, but anyway, we need to follow Babel's implementation 😢

overlookmotel replied:
We need to follow Babel's *output*, but I don't think we need follow its *implementation* necessarily.

Anyway, I don't fully understand this transform, so I may be wrong in what I said above anyway.

overlookmotel reviewed (approved):
I've not reviewed the whole transform, but nice one getting the semantic tests to pass!

overlookmotel commented:
### Merge activity

* **Oct 30, 8:03 AM EDT**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

overlookmotel commented:
@Dunqing Looks like CI has got stuck. I see it do this sometimes, don't know why. Solution I've found is:

* Make a random change (add a dummy comment, whatever) and `gt m -a`.
* Revert that change and `gt m -a` again.
* `gt submit`

Or if branch has got behind main, then just `gt sync` + `gt submit` also works.

Dunqing commented:
> @Dunqing Looks like CI has got stuck. I see it do this sometimes, don't know why. Solution I've found is:
> 
> * Make a random change (add a dummy comment, whatever) and `gt m -a`.
> * Revert that change and `gt m -a` again.
> * `gt submit`
> 
> Or if branch has got behind main, then just `gt sync` + `gt submit` also works.

Thanks! That works, just an empty commit so I think it's fine if we don't revert it

Dunqing committed:
feat(transformer): add `async_generator_functions` plugin (#6573)

Passed 15/19 tests. The remaining 4 failed tests related to `this` expression, the problem same as I mentioned in #6658. I will fix them in follow-up PRs.
